app/bot/cogs/economics.py

```diff
@@ -0,0 +1,110 @@
+import math
+import typing
+
+import discord
+from discord.ext import commands
+
+from app.db.usermodel import SqlUser
+from app.helper_tools import basic_embed
+from app.entities.users import User
+from app.config import logovo_config
+from app.db import database
+
+
+class EconomicsCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot = bot
+
+    @discord.app_commands.rename(duser="цель")
+    @commands.hybrid_command(
+        name="карма",
+        description="посмотреть свою или чью-то карму."
+    )
+    async def view_karma(self, ctx, duser: typing.Optional[discord.Member]):
+        # if len(args) == 1:
+        #     user = User(msg.author.id)
+        # else:
+        #     user = User.from_string(args[1])
+        #     if not user:
+        #         embed = basic_embed(":x: Ээээ....", "кто?")
+        #         embed.color = discord.Color.red()
+        #         await msg.channel.send(embed=embed)
+        #         return
+        if not duser:
+            duser = ctx.author
+        user = User(duser.id)
+        embed = basic_embed(
+            "Профиль " + duser.name, "Постовая карма: " + str(user.karma)
+        )
+        if duser.avatar:
+            embed.set_thumbnail(url=duser.avatar.url)
+        await ctx.send(embed=embed)
+
+
+    @discord.app_commands.rename(duser="номер страницы")
+    @commands.hybrid_command(
+        name="лидеры",
+        description="топ кармов."
+    )
+    async def leaderboard(self, ctx, page: int = 1):
+        db_sess = database.session()
+        users = (
+            db_sess.query(SqlUser)
+            .filter(SqlUser.karma != 0)
+            .order_by(SqlUser.karma.desc())
+        )
+        maxpage = math.ceil(users.count() / 10)
+        page = max(1, min(maxpage, page))
+        text = ""
+        for i, user in enumerate(users[(page - 1) * 10 : page * 10]):
+            text += (
+                "**"
+                + str((page - 1) * 10 + i + 1)
+                + ".** `["
+                + str(user.karma)
+                + "]` <@!"
+                + str(user.discord_id)
+                + ">\n"
+            )
+        embed = basic_embed(
+            title="Топ кармов (Страница " + str(page) + "/" + str(maxpage) + ")",
+            text=text,
+        )
+
+        await ctx.send(embed=embed)
+
+    async def reaction_event(self, payload, meaning=1):
```

Comment:
On one hand, it makes total sense, that Karma calculation functionality is implemented in the `economics.py` module. On the other hand, it seems counterintuitive, that on_raw_reaction event handlers are written inside `/cogs/`. I may be wrong, but what do these handlers have to do with cogs? I want to discuss this a little bit

Reply:
i think about cogs as of _groups of commands and listeners targeted at a specific functionality_. in this case, economics. separating listeners in cogs also allows to create _multiple listeners to one event_, which is more convenient, than writing multiple separate logics in one single listener.

also, if you're working, for example, with economics, you don't need to jump between several files, in order to review it's commands and listeners, which is pretty good in my opinion

Reply:
hm, fair enough. I'll keep that